### DISCUSSION TITLE
Additionnal trust via CA Bundle not working at all ?

### DISCUSSION THREAD

**Discussion by RyuunoAelia** [okd-project/okd#1291] [General]
**Describe the bug**
I am trying to get OKD 4.10.0-0.okd-2022-07-09-073606 to download an image from a gitlab registry hosted inside openshift, however it fails to validate the certificate whatever I configure in image.config.openshift.io/cluster and proxy.config.openshift.io/cluster

In the proxy configuration I have:
```
spec:
  trustedCA:
    name: user-ca-bundle
```

And the ConfigMap:
```
$ oc get cm -n openshift-config user-ca-bundle -o yaml
apiVersion: v1
data:
  ca-bundle.crt: |
    -----BEGIN CERTIFICATE-----
    MIIDjjCCAnagAwIBAgIBATANBgkqhkiG9w0BAQsFADA3MRUwEwYDVQQKDAxJRE0u
    QUVMSUEuQ0gxHjAcBgNVBAMMFUNlcnRpZmljYXRlIEF1dGhvcml0eTAeFw0xOTAy
    MTYxMzI0MjFaFw0zOTAyMTYxMzI0MjFaMDcxFTATBgNVBAoMDElETS5BRUxJQS5D
    SDEeMBwGA1UEAwwVQ2VydGlmaWNhdGUgQXV0aG9yaXR5MIIBIjANBgkqhkiG9w0B
    AQEFAAOCAQ8AMIIBCgKCAQEArOGi8WpYgO6oFJAQHyTkMTiOa3zTpj/+XMPkSBQ2
    FMZzEiHjCYpmyJaPVPNB7EeeqLt60K1FDNFISXOoa+R6TcLmMVTPyd71Zuea3pVc
    ubiL/mZIOa7T403HeS1A9i9TnryrqUtvdnkk+hoY5zWjuaQ1cUr4ftiQ8Jky7TQS
    rc83MgYlBq0nqBvCCBi41Hk/0qMNMnvZG8gb9Yazr+iKnUFxEm277DGPzeUIsEFR
    2SXpXYAbEk+5wqXDslLpvdP9rG/NPsfYcGD760jxlObuBy1gpa5t8zg9oIHb2B4J
    TmcKr/h9+Skv6TgCx+3xrobG9uJU5AnU/QL/N8Ck1Yyv3wIDAQABo4GkMIGhMB8G
    A1UdIwQYMBaAFFGI+b6CHmm3iTK2UjROklL7oeaMMA8GA1UdEwEB/wQFMAMBAf8w
    DgYDVR0PAQH/BAQDAgHGMB0GA1UdDgQWBBRRiPm+gh5pt4kytlI0TpJS+6HmjDA+
    BggrBgEFBQcBAQQyMDAwLgYIKwYBBQUHMAGGImh0dHA6Ly9pcGEtY2EuaWRtLmFl
    bGlhLmNoL2NhL29jc3AwDQYJKoZIhvcNAQELBQADggEBADudi0zWz3zUys3cTHep
    ddwOkrOljxlxgnjYpZwKT+104OHvW1Li1WkvOvJNSwJkFRuEmM/iISphn8SU6ihY
    SkCzsUiG7Aw0nT/5G1y9dQ1k5SGYNlOOI4vwJs+ARY7Hej+0TzWvbfgheXZxa8qL
    KD/popxLzmTSLFt6xj2K5uhCUftZkf7UfuornoznemmrEJkZfZUINxDWXwa2W7YW
    Dtp2GD2kevyxFQfiOhALp/Kb3MJVWjXV6iHFqMu9dq5drEO20BfD02KOHQbIzl0R
    PQB27NAQ8RzY60EJjiMpFpdffzuv+qOUGZHI9CsuK/repoEAyWUQiBJOLgO1KKZb
    bGQ=
    -----END CERTIFICATE-----
kind: ConfigMap
metadata:
  creationTimestamp: "2020-05-26T11:41:20Z"
  name: user-ca-bundle
  namespace: openshift-config
  resourceVersion: "5948088"
  uid: 182f44fd-edc0-4baf-bb54-f20fae120184
```

This is indeed deployed on the coreos machines in `/etc/pki/ca-trust/source/anchors/openshift-config-user-ca-bundle.crt` but update-ca-trust is not executed so the anchors are not placed into the system trust bundle... so even a curl will fail.

However executing the `update-ca-trust` command is not enough, I tried to put the CA in the image configuration, but nothing happens:
```
apiVersion: config.openshift.io/v1
kind: Image
[...]
spec:
  additionalTrustedCA:
    name: registry-ca-bundle
```
and the ConfigMap:
```
oc get cm -n openshift-config registry-ca-bundle -o yaml
apiVersion: v1
data:
  registry.apps.okd4.os.aelia.ch: |
    -----BEGIN CERTIFICATE-----
    MIIDjjCCAnagAwIBAgIBATANBgkqhkiG9w0BAQsFADA3MRUwEwYDVQQKDAxJRE0u
    QUVMSUEuQ0gxHjAcBgNVBAMMFUNlcnRpZmljYXRlIEF1dGhvcml0eTAeFw0xOTAy
    MTYxMzI0MjFaFw0zOTAyMTYxMzI0MjFaMDcxFTATBgNVBAoMDElETS5BRUxJQS5D
    SDEeMBwGA1UEAwwVQ2VydGlmaWNhdGUgQXV0aG9yaXR5MIIBIjANBgkqhkiG9w0B
    AQEFAAOCAQ8AMIIBCgKCAQEArOGi8WpYgO6oFJAQHyTkMTiOa3zTpj/+XMPkSBQ2
    FMZzEiHjCYpmyJaPVPNB7EeeqLt60K1FDNFISXOoa+R6TcLmMVTPyd71Zuea3pVc
    ubiL/mZIOa7T403HeS1A9i9TnryrqUtvdnkk+hoY5zWjuaQ1cUr4ftiQ8Jky7TQS
    rc83MgYlBq0nqBvCCBi41Hk/0qMNMnvZG8gb9Yazr+iKnUFxEm277DGPzeUIsEFR
    2SXpXYAbEk+5wqXDslLpvdP9rG/NPsfYcGD760jxlObuBy1gpa5t8zg9oIHb2B4J
    TmcKr/h9+Skv6TgCx+3xrobG9uJU5AnU/QL/N8Ck1Yyv3wIDAQABo4GkMIGhMB8G
    A1UdIwQYMBaAFFGI+b6CHmm3iTK2UjROklL7oeaMMA8GA1UdEwEB/wQFMAMBAf8w
    DgYDVR0PAQH/BAQDAgHGMB0GA1UdDgQWBBRRiPm+gh5pt4kytlI0TpJS+6HmjDA+
    BggrBgEFBQcBAQQyMDAwLgYIKwYBBQUHMAGGImh0dHA6Ly9pcGEtY2EuaWRtLmFl
    bGlhLmNoL2NhL29jc3AwDQYJKoZIhvcNAQELBQADggEBADudi0zWz3zUys3cTHep
    ddwOkrOljxlxgnjYpZwKT+104OHvW1Li1WkvOvJNSwJkFRuEmM/iISphn8SU6ihY
    SkCzsUiG7Aw0nT/5G1y9dQ1k5SGYNlOOI4vwJs+ARY7Hej+0TzWvbfgheXZxa8qL
    KD/popxLzmTSLFt6xj2K5uhCUftZkf7UfuornoznemmrEJkZfZUINxDWXwa2W7YW
    Dtp2GD2kevyxFQfiOhALp/Kb3MJVWjXV6iHFqMu9dq5drEO20BfD02KOHQbIzl0R
    PQB27NAQ8RzY60EJjiMpFpdffzuv+qOUGZHI9CsuK/repoEAyWUQiBJOLgO1KKZb
    bGQ=
    -----END CERTIFICATE-----
kind: ConfigMap
metadata:
  creationTimestamp: "2022-07-30T09:21:01Z"
  name: registry-ca-bundle
  namespace: openshift-config
  resourceVersion: "588996071"
  uid: 351760e0-6e54-49aa-8519-218d7c2adfc0
```

My understanding is that this should be taken by MCO and put into MachineConfig `99-worker-generated-registries` and `99-master-generated-registries` however these two objects are not updated to reflect the change I made and are completely empty.

**Version**
UPI 4.10.0-0.okd-2022-07-09-073606

**How reproducible**
Try to configure an Additionnal CA Bundle, and coreos is not using the new CA.

**Log bundle**
I looked everywhere and saw no logs at all concerning this issue except the pull errors saying that the CA cannot be validated.
If you still want the logs I can prepare a google drive link.

**Comment by vrutkovs** [comment]
>In the proxy configuration I have:
>spec:
>  trustedCA:
>   name: user-ca-bundle

This will only affect proxy configuration, not crio pulls

> tried to put the CA in the image configuration, but nothing happens:
>apiVersion: config.openshift.io/v1
>kind: Image
>[...]
>spec:
> additionalTrustedCA:
>    name: registry-ca-bundle

This will affect image registry imports, not crio pulls.

See https://docs.openshift.com/container-platform/4.10/openshift_images/managing_images/using-image-pull-secrets.html#images-allow-pods-to-reference-images-from-secure-registries_using-image-pull-secrets

**Reply by QkiZMR** [reply]
your link is about adding image registry credentials, not adding custom ca to okd4

**Comment by RyuunoAelia** [comment]
@vrutkovs I installed literally dozens of OCP4 clusters in enterprise setups where they used proxies and private registries so I consider myself knowing pretty well what settings should be set.
The dockerconfig.json setup if for authentication of the Registry it does not contain the CA certificate to validate the trust chain of the registry.
The proper place to set the CA for a specific registry should be in `/etc/docker/certs.d` on the  host... however from what I see in my setup, the file is there and on the coreos machine when I perform podman actions it is taken, however when kubernetes tries to pull  images it, it says the CA is not valid...

This is why I opened a bug it seems something is not working as expected (like the update-ca-trust for the system-wide trust store).
I will try to reboot one machine and see if this changes anything.

**Reply by vrutkovs** [reply]
>The proper place to set the CA for a specific registry should be in /etc/docker/certs.d on the host

Why? OKD 4 doesn't use docker

>like the update-ca-trust for the system-wide trust store

It works here on F35, so why wouldn't it work on OKD OS? Does curl work after `update-ca-trust extract`?

**Reply by RyuunoAelia** [reply]
Ok sorry for the noise, the issue was not with openshift, it was the registry... to be clear the registry was redirecting the client (CRI-o) to download the blobs directly from the Object Storage... And this was where the errors about certificate validation failed, the error message was writing that the validation of the certificate of the registry (with the explicit hostname of the registry) was not working but it was working. This confused me and made me search at the wrong spot.
And I tried logging into the registry with podman and pushing into the registry with kaniko which was working fine and this made me even more confused...

I reconfigured the registry to proxy all access and now it works.

**Reply by RyuunoAelia** [reply]
As for the location of the certificates, I have no idea why they made it /etc/docker/certs.d but this is exactly where openshift puts them when you add them:
```
[root@okd4-compute-201 core]# ls -alhR /etc/docker/certs.d/
/etc/docker/certs.d/:
total 0
drwxr-xr-x. 5 root root 170 Jul 30 09:23 .
drwxr-xr-x. 3 root root  37 Jul 25 08:21 ..
drwxr-xr-x. 2 1001 root  20 Jul 25 08:21 image-registry.openshift-image-registry.svc.cluster.local:5000
drwxr-xr-x. 2 1001 root  20 Jul 25 08:21 image-registry.openshift-image-registry.svc:5000
drwxr-xr-x. 2 1001 root  20 Jul 30 09:23 registry.apps.okd4.os.aelia.ch

'/etc/docker/certs.d/image-registry.openshift-image-registry.svc.cluster.local:5000':
total 4.0K
drwxr-xr-x. 2 1001 root   20 Jul 25 08:21 .
drwxr-xr-x. 5 root root  170 Jul 30 09:23 ..
-rw-r--r--. 1 1001 root 2.4K Jul 30 09:23 ca.crt

'/etc/docker/certs.d/image-registry.openshift-image-registry.svc:5000':
total 4.0K
drwxr-xr-x. 2 1001 root   20 Jul 25 08:21 .
drwxr-xr-x. 5 root root  170 Jul 30 09:23 ..
-rw-r--r--. 1 1001 root 2.4K Jul 30 09:23 ca.crt

/etc/docker/certs.d/registry.apps.okd4.os.aelia.ch:
total 4.0K
drwxr-xr-x. 2 1001 root   20 Jul 30 09:23 .
drwxr-xr-x. 5 root root  170 Jul 30 09:23 ..
-rw-r--r--. 1 1001 root 1.3K Jul 30 09:23 ca.crt
```

Though I have no idea what component does that since it is not in the MachineConfig rendered manifest of the machines..

**Comment by RyuunoAelia** [comment]
Rebooting the CoreOS machine did not change anything. CRI-o still does not take the CA certificate in `/etc/docker/certs.d` while podman does work.
